cpp/ql/src/experimental/Security/CWE/CWE-266/IncorrectPrivilegeAssignment.ql

```diff
@@ -0,0 +1,77 @@
+/**
+ * @name Find the wrong use of the umask function.
+ * @description Incorrectly evaluated argument to the umask function may have security implications.
+ * @kind problem
+ * @id cpp/wrong-use-of-the-umask
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       maintainability
+ *       security
+ *       external/cwe/cwe-266
+ *       external/cwe/cwe-264
+ *       external/cwe/cwe-200
+ *       external/cwe/cwe-560
+ *       external/cwe/cwe-687
+ */
+
+import cpp
+import semmle.code.cpp.exprs.BitwiseOperation
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds for a function `f` that has an argument at index `apos` used to set file permissions. */
+predicate numberArgumentModFunctions(Function f, int apos) {
+  f.hasGlobalOrStdName("umask") and apos = 0
+  or
+  f.hasGlobalOrStdName("fchmod") and apos = 1
+  or
+  f.hasGlobalOrStdName("chmod") and apos = 1
+}
+
+from FunctionCall fc, string msg
+where
+  fc.getTarget().hasGlobalOrStdName("umask") and
+  fc.getArgument(0).getValue() = "0" and
+  not exists(FunctionCall fctmp |
+    fctmp.getTarget().hasGlobalOrStdName("umask") and
+    globalValueNumber(fctmp.getArgument(0)) != globalValueNumber(fc.getArgument(0))
+  ) and
+  exists(FunctionCall fctmp |
+    (
+      fctmp.getTarget().hasGlobalOrStdName("fopen") or
+      fctmp.getTarget().hasGlobalOrStdName("open")
+    ) and
+    fctmp.getNumberOfArguments() = 2 and
+    fctmp.getArgument(0).getValue() != "/dev/null"
+  ) and
+  not exists(FunctionCall fctmp |
+    fctmp.getTarget().hasGlobalOrStdName("chmod") or
+    fctmp.getTarget().hasGlobalOrStdName("fchmod")
+  ) and
+  msg = "Using umask (0) may not be safe."
+  or
+  fc.getTarget().hasGlobalOrStdName("umask") and
+  exists(FunctionCall fctmp |
+    (
+      fctmp.getTarget().hasGlobalOrStdName("chmod") or
+      fctmp.getTarget().hasGlobalOrStdName("fchmod")
+    ) and
+    (
+      globalValueNumber(fc.getArgument(0)) = globalValueNumber(fctmp.getArgument(1)) and
+      fc.getArgument(0).getValue() != "0"
+    ) and
+    msg = "not use equal argument in umask and " + fctmp.getTarget().getName() + " functions"
```

Comment:
I'm not convinced this case is right:
 - if the `umask` and the `chmod` values are equal, the resulting file permissions are `0`.  This may or may not be what the programmer intended, but `0` is a very safe permission setting.
    - I think what we really want is that the argument to `chmod` minus the bits from `umask` contains at least one bit we consider dangerous, e.g. global write permission, suggesting a dangerous setup.
 - if there's more than one call to `umask` in the program, we don't know which `chmod` calls correspond with which `umask` call.  I'm not sure how common it is in practice to set `umask` anywhere other than once at the beginning of the program.

Reply:
I agree with what you say. (except for using 0 in `chmod`, I can't figure out why disallow everything for my file)

in the request, I wanted to take a broader look at the cwe-560. and for me the error indicator is the use of the same value in opposite functions. in my opinion, this is a good criterion, devoid of false detection.

but if you insist, I will remove this section from the request

Reply:
I see what you mean, even if a resulting permission of `0` is safe, it may indicate a mistake by the programmer.  Perhaps we should wait until we run the query on LGTM (after resolving the other questions) and see whether your approach produces good results in practice?